### DISCUSSION TITLE
How to export the certificate private key from the end entity ?

### DISCUSSION THREAD

**Discussion by thomascenni** [Keyfactor/ejbca-ce#583] [Q&A]
Hello,
I have installed EJBCA CE version 8.2.0.1. I have created some end entity, and I need the PEM file and the KEY file to connect to a MQTT instance. The PEM file is available directly on the certificate page, how to export the KEY file ?

![image](https://github.com/Keyfactor/ejbca-ce/assets/33039107/fc44bfa7-7947-434f-a794-45718d880747)

**Comment by primetomas** [comment]
This all depends how you have enrolled for the certificate.
Private keys are by default not stored by the CA, which is best practice for security reasons. If you used server generated keys, with [key recovery](https://doc.primekey.com/ejbca/ejbca-operations/ejbca-operations-guide/ca-operations-guide/key-recovery) enabled, you could perform a key recovery operation and recover the private key.

If you uploaded a CSR to get the certificate, the CA has never seen the private key, and it was generated by your client, and you should get the file from where you generated the CSR.

If you used "PEM" keystore type when issuing the certificate (but not key recovery enabled), the private key was generated by the CA and sent to the client, but not stored on the CA side. Same there, you can get the private key from the file you downloaded when issuing the certificate.

Here's another quick start guide for issuing certificates. There are many different ways to do this, in order to fit many different use cases.
https://doc.primekey.com/ejbca/tutorials-and-guides/quick-start-guide-issue-client-authentication-certificate-using-ejbca

**Reply by thomascenni** [reply]
Hi and thanks for the rapid answer!
I have "Enable Key Recovery" active on "System Configuration" and set the checkbox "Key Recoverable" when I make the request:

![image](https://github.com/Keyfactor/ejbca-ce/assets/33039107/d8acbdde-cf69-41e7-b3d8-19cd3011566e)

But then I am not able to perform the key recovery operation you mention.

**Reply by primetomas** [reply]
If the enrollment was succesful, with a server generated keystore, and no error happened when storung the key recovery data, you should get a button to the left when viewing the certificate "Recover Key"
![Screenshot from 2024-05-02 17-01-28](https://github.com/Keyfactor/ejbca-ce/assets/28708578/0ada2dc3-310e-436c-aba6-44633378cdf6)

**Reply by primetomas** [reply]
Check the server.log file when enrolling, the audit log will say when key recovery data is store.

**Reply by thomascenni** [reply]
Still not able to get the key. I'll start with a fresh installation and follow the tutorial you mentioned before.
